Brokerages/InteractiveBrokers/InteractiveBrokersBrokerage.cs

```diff
@@ -1388,18 +1430,84 @@ private void HandleExecutionDetails(object sender, IB.ExecutionDetailsEventArgs
+                {
+                    // save commission in dictionary and wait for execution event
+                    _commissionReports[e.CommissionReport.ExecId] = e.CommissionReport;
+                    return;
+                }
+
+                var order = _orderProvider.GetOrderByBrokerageId(execution.OrderId);
+                if (order == null)
+                {
+                    Log.Error("InteractiveBrokersBrokerage.HandleExecutionDetails(): Unable to locate order with BrokerageID " + execution.OrderId);
+                    return;
+                }
+
+                if (CanEmitFill(order, execution))
+                {
+                    // we have both execution and commission report, emit the fill
+                    EmitOrderFill(order, execution, e.CommissionReport);
+                }
+
+                // always remove previous execution
+                _orderExecutions.TryRemove(e.CommissionReport.ExecId, out execution);
+            }
+            catch (Exception err)
             {
-                var security = _securityProvider.GetSecurity(order.Symbol);
-                orderFee = security.FeeModel.GetOrderFee(security, order);
+                Log.Error("InteractiveBrokersBrokerage.HandleCommissionReport(): " + err);
             }
+        }
+
+        /// <summary>
+        /// Decide which fills should be emitted, accounting for different types of Financial Advisor orders
+        /// </summary>
+        private bool CanEmitFill(Order order, Execution execution)
+        {
+            if (order.Status == OrderStatus.Filled)
+                return false;
+
+            // non-FA orders
+            if (!IsFinancialAdvisor)
+                return true;
+
+            var orderProperties = order.Properties as InteractiveBrokersOrderProperties;
+            if (orderProperties == null)
+                return true;
+
+            return
+                // FA master orders for groups/profiles
+                string.IsNullOrWhiteSpace(orderProperties.Account) && execution.AcctNumber == _account ||
+
+                // FA orders for single managed accounts
+                !string.IsNullOrWhiteSpace(orderProperties.Account) && execution.AcctNumber == orderProperties.Account;
+        }
+
+        /// <summary>
+        /// Emits an order fill (or partial fill) including the actual IB commission paid
+        /// </summary>
+        private void EmitOrderFill(Order order, Execution execution, CommissionReport commissionReport)
+        {
+            var currentQuantityFilled = Convert.ToInt32(execution.Shares);
+            var totalQuantityFilled = Convert.ToInt32(execution.CumQty);
+            var remainingQuantity = Convert.ToInt32(order.AbsoluteQuantity - totalQuantityFilled);
+            var price = Convert.ToDecimal(execution.Price);
+            var orderFee = Convert.ToDecimal(commissionReport.Commission);
```

Comment:
Wasn't there something about partial fills leading to duplicate fees or was that for a different brokerage impl?

Reply:
Yes, the issue is in the GDAX brokerage (work in progress)